### PR TITLE
la

### PR DESCRIPTION
**Pull request by LiliyanAV**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    def crypt_chr(self, ch):
+       return chr((ord(ch) - ord('a') + self.shift) % self.size + ord('a'))
+
+    def crypt(self, text):
+        text.lower()
+        return "".join([self.crypt_chr(chr) for chr in text])
+
+    def encrypt(self, encoded_text):
+        return Caesar.crypt(self, encoded_text)
```
Comment (HarIgo23) [inline main.py:14]: You should use self, not class name. Rewrite

File: main.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class Caesar:
+    def __init__(self):
+        self.size = ord('z') - ord('a')+1
```
Comment (HarIgo23) [inline main.py:3]: You have to add space between operator

File: main.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        self.shift = 13
+
+    def crypt_chr(self, ch):
+       return chr((ord(ch) - ord('a') + self.shift) % self.size + ord('a'))
+
+    def crypt(self, text):
+        text.lower()
+        return "".join([self.crypt_chr(chr) for chr in text])
```
Comment (HarIgo23) [inline main.py:11]: You don't handle special symbols, spaces and snother.

File: test.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import unittest
```
Comment (HarIgo23) [inline test.py:1]: Use pytest for testing

File: test.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+from main import Caesar
+
+
+class TestCaesaer(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.p = Caesar()
+
+    def test_crypt(self):
+        self.assertEqual(self.p.crypt("aha"), "nun")
```
Comment (HarIgo23) [inline test.py:11]: There are useless test. You should check all alphabet in asc and desc order + maybe some random phrase with special symbols, space and uppercase letter.

**Review by HarIgo23 — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v2.3.0
+    hooks:
+      - id: check-yaml
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+  - repo: https://github.com/psf/black
+    rev: 19.3b0
+    hooks:
```
Comment (HarIgo23) [inline .pre-commit-config.yaml:10]: add flake8 to precommit

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+/python-vitual-environments/
```
Comment (HarIgo23) [inline .gitignore:1]: Add venv/ and env/ in .gitignore

File: main.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+class Caesar:
+    def __init__(self):
```
Comment (HarIgo23) [inline main.py:2]: Use shift  argument with default value as 13

File: main.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+        self.shift = 13
```
Comment (HarIgo23) [inline main.py:3]: Use argument to set up value

File: main.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+        self.alpha = [chr(x) for x in range(ord('a'), ord('z') + 1)]
```
Comment (HarIgo23) [inline main.py:4]: alphabet

File: main.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        self.size = ord('z') - ord('a') + 1
```
Comment (HarIgo23) [inline main.py:5]: len(self.alphabet)

File: main.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+    def crypt_chr(self, ch):
+       return chr((ord(ch) - ord('a') + self.shift) % self.size + ord('a')) if ch in self.alpha else ch
```
Comment (HarIgo23) [inline main.py:8]: Rewrite as
`
if ch in self.alpha:
    return chr((ord(ch) - ord('a') + self.shift) % self.size + ord('a'))
return ch 
`
, it's too complicated expression for one line

File: requirements.txt [hunk 13]
```diff
@@ -0,0 +1,8 @@
+attrs==21.2.0
```
Comment (HarIgo23) [inline requirements.txt:1]: Where are pre-commit and pytest?

File: test_main.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import pytest
+from main import Caesar
```
Comment (HarIgo23) [inline test_main.py:2]: https://www.python.org/dev/peps/pep-0008/#imports
Imports should be grouped in the following order:

1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.

Comment (HarIgo23) [reply test_main.py:2]: '
import pytest

from main import Caesar
'
You should write it this way, bcs pytest - third party imports, main - local application imports.

File: test_main.py [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+
+
+def test_crypt():
+    c = Caesar()
+    assert c.crypt("b arf") == "o nes"
+    assert c.crypt("b arf") == "o nes"
```
Comment (HarIgo23) [inline test_main.py:8]: Add more test with cpecific symbols.

**Review by HarIgo23 — CHANGES_REQUESTED** [review]
You forgot to add Readme file you write instructions how set up you project on computer.

File: .gitignore [hunk 16]
```diff
@@ -0,0 +1,3 @@
+/env/
+
+/env/env/
```
Comment (HarIgo23) [inline .gitignore:3]: What is it?

File: README.md [hunk 17]
```diff
@@ -1 +1,59 @@
...
+
+------------
+
+This project requires at least python 3.9 and other.
+
+> :arrow_right: mkdir caesar
+
+> :arrow_right: cd caesar
+
+> :arrow_right: git clone https://github.com/Luchia04/caesar.git
```
Comment (HarIgo23) [inline README.md:23]: Change to
git clone https://github.com/Luchia04/caesar.git caesar
cd caesar

and delete first two command where you create and move into directory

File: README.md [hunk 18]
```diff
@@ -1 +1,59 @@
...
+
+### Install requirements:
+
+------------
+
+> :arrow_right: python3 -m venv .venv
+
+> :arrow_right: source .venv/bin/activate
+
+> :arrow_right: pip install --upgrade pip
```
Comment (HarIgo23) [inline README.md:37]: No need this command

File: README.md [hunk 19]
```diff
@@ -1 +1,59 @@
...
+
+------------
+
+> :arrow_right: pre-commit install
+
+### Run flake8
+
+----------
+
+> :arrow_right: flake8 main.py
```
Comment (HarIgo23) [inline README.md:59]: Where chapter with instruction how run your app?

File: main.py [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+        self.shift = shift
+        self.alphabet = [chr(x) for x in range(ord("a"), ord("z") + 1)]
+        self.size = len(self.alphabet)
+
+    def crypt_chr(self, ch):
+        if ch not in self.alphabet:
+            return ch
+        return chr((ord(ch) - ord("a") + self.shift) % self.size + ord("a"))
+
+    #  return (
```
Comment (HarIgo23) [inline main.py:12]: Delete unused code

File: requirements.txt [hunk 21]
```diff
@@ -0,0 +1,10 @@
...
```
Comment (HarIgo23) [inline requirements.txt:1]: Use pip freeze > requirements.txt to write into file all dependences

File: test_main.py (hunk 15 above)
Comment (HarIgo23) [reply test_main.py:7]: You didn't resolve this problem. You need add more TEST CASES in TESTS file.

File: test_main.py (hunk 14 above)
Comment (HarIgo23) [reply test_main.py:1]: Why you remove pytest. You need change order. NOT delete this import.

**Review by HarIgo23 — CHANGES_REQUESTED** [review]

File: test_main.py (hunk 14 above)
Comment (Luchia04) [reply test_main.py:1]: I can't pushing this order test_main.py:2:1: F401 'pytest' imported but unused

File: .gitignore (hunk 16 above)
Comment (Luchia04) [reply .gitignore:3]: It is .gitignore

**Review by HarIgo23 — CHANGES_REQUESTED** [review]

**Review by HarIgo23 — APPROVED** [review]

**Pull request merged**